scipy/stats/_continuous_distns.py

```diff
@@ -3554,6 +3559,66 @@ def fit(self, data, *args, **kwds):
             fshape_s = fshape_n / fscale
         return fshape_s, floc, fscale
 
+    def _ppf_isf(self, q, mu, upper):
+        """ calculate ppf or isf depending on the value of `upper`.
+        If upper is True, then compute isf, else ppf. This implementation is
+        based on Gyner & Smyth (2016) and uses newton's method with carefully
+        selected starting points that guarantee convergence regardless of the
+        probability."""
+
+        def f(a, b, c, d, e):
+            """approximates (x0 - norm_cdf) / pdf when abs(x0 - norm_cdf) < 1e-5"""
+            return a * np.exp(c + np.log1p(-0.5 * a) - d)
+
+        def f2(a, b, c, d, e):
+            """computes (x0 - norm_cdf) / pdf"""
+            return b * np.exp(-d) - np.exp(e - d)
+
+        def f_small_q(q, mu, k):
+            """select the start values for the newton using the quantiles of
+            the gamma (if right tail is required) or normal distribution"""
+            if upper:
+                mu3 = mu ** 3
+                return sc.gammainccinv(1 / mu3, q) / mu3
+            return mu / (_norm_ppf(q) ** 2),
+
+        k = 1.5 * mu
+        # get starting points
+        x0 = _lazywhere(
```

Comment:
Both branches are complicated conditions here. So an explicit array indexing and separate branch computations would make this code much more readable in my opinion. Now it's a bit of lambda olympics and not very readable

Reply:
I've been trying this out for the last few minutes. It doesn't seem to be working out well for me because `q`, `mu` can be either scalars or arrays or both. So having to handle that outside of this _lazywhere makes the code much more verbose and unreadable that it was before. See below:
```python
        q = np.asarray(q)
        mu = np.asarray(mu)
		...
        # get starting values to use with Newton's Method
        k = 1.5 * mu
        x0 = np.empty_like(q if q.size > mu.size else mu)
        mask = q < 1e-05
        if mask.any():
            # see page 8 of Giner & Smyth (2016)
            x0[mask] = small_q_init_value(q[mask], mu[mask], k[mask])
        # when q is not very small the initial value for Newton's method is
        # the mode of the distribution. If `k` is large, then we use equation 3
        # of Giner & Smyth (2016). Note that the threshold of 100 picked here
        # is arbitrary, a much bigger value could be chosen if necessary.
        x0[~mask] = _lazywhere(
            mu[~mask] > 100,
            (mu[~mask], k[~mask]),
            lambda mu, k: mu * (0.5 / k - 0.125 / (k ** 3) + 0.0625 / (k ** 6)),
            f2=lambda mu, k: mu * (np.sqrt(1 + k * k) - k)
        )
```
any suggestions? Tests still fail locally when I re-write this way due to incompatible input sizes.